src/hotspot/share/opto/mulnode.cpp

```diff
@@ -875,26 +889,28 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
 //=============================================================================
 //------------------------------Identity---------------------------------------
 Node* RShiftINode::Identity(PhaseGVN* phase) {
-  int shift = getShiftCon(phase, this, -1);
-  if (shift == -1) return this;
-  if ((shift & (BitsPerJavaInteger - 1)) == 0) return in(1);
-
-  // Check for useless sign-masking
-  if (in(1)->Opcode() == Op_LShiftI &&
-      in(1)->req() == 3 &&
-      in(1)->in(2) == in(2)) {
-    shift &= BitsPerJavaInteger-1; // semantics of Java shifts
-    // Compute masks for which this shifting doesn't change
-    int lo = (-1 << (BitsPerJavaInteger - ((uint)shift)-1)); // FFFF8000
-    int hi = ~lo;               // 00007FFF
-    const TypeInt *t11 = phase->type(in(1)->in(1))->isa_int();
-    if (!t11) return this;
-    // Does actual value fit inside of mask?
-    if (lo <= t11->_lo && t11->_hi <= hi) {
-      return in(1)->in(1);      // Then shifting is a nop
+  int count = 0;
+  if (const_shift_count(phase, this, &count)) {
+    if ((count & (BitsPerJavaInteger - 1)) == 0) {
+      // Shift by a multiple of 32 does nothing
+      return in(1);
+    }
+    // Check for useless sign-masking
+    if (in(1)->Opcode() == Op_LShiftI &&
+        in(1)->req() == 3 &&
+        in(1)->in(2) == in(2)) {
+      count &= BitsPerJavaInteger-1; // semantics of Java shifts
+      // Compute masks for which this shifting doesn't change
+      int lo = (-1 << (BitsPerJavaInteger - ((uint)count)-1)); // FFFF8000
+      int hi = ~lo;               // 00007FFF
+      const TypeInt *t11 = phase->type(in(1)->in(1))->isa_int();
+      if (!t11) return this;
```

Comment:
While at it, you could also improve the styling here (spacing, asterisk and explicitly checking against `NULL` for `t11`).